[FLINK-25085][runtime] Add a scheduled thread pool for periodic tasks in RpcEndpoint

zjureel opened the pull request:
## What is the purpose of the change
This PR adds a scheduled thread pool which will used to hold the periodic tasks and send them to akka with gateway when they should be executed.  The thread pool will be shutdown when the RpcEndpoint is stopped.

## Brief change log

  -* Add a scheduled thread pool to hold the periodic tasks and send them to akka with gateway when they should be executed
  -* Add closeable resource registry in RpcEndpoint and close all the resources which are registered in it.
  -* The thread pool in RpcEndpoint will register itself to the closeable resource registry

## Verifying this change

This change added tests and can be verified as follows:
  -* Add PeriodicTaskScheduledExecutorTest to test schedule runnable and callable task in the thread pool
  -* Add validation in JobMasterTest to check if the thread pool is closed

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no) no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no) no
  - The serializers: (yes / no / don't know) no
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know) no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know) no
  - The S3 file system connector: (yes / no / don't know) no

## Documentation

  - Does this pull request introduce a new feature? (yes / no) no
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

zjureel commented:
@flinkbot run azure

zjureel commented:
@flinkbot run azure

zjureel commented:
@KarmaGYZ I create a new PR for FLINK-25085 to add a thread pool in RpcEndpoint for periodic tasks and I will run the ci for several times. Could you please review it first? thanks :)

KarmaGYZ commented on flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/concurrent/PeriodicTaskScheduledExecutor.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    @Override
+    public void execute(Runnable command) {
```
```suggestion
    public void execute(@Nonnull Runnable command) {
```

KarmaGYZ commented on flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/concurrent/ThrowingScheduledFuture.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    @Override
+    public T get(long timeout, TimeUnit unit)
```
```suggestion
    public T get(long timeout, @Nonnull TimeUnit unit)
```

KarmaGYZ commented on flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/concurrent/ThrowingScheduledFuture.java:
```diff
@@ -0,0 +1,76 @@
+    private ThrowingScheduledFuture() {}
+
+    @Override
+    public long getDelay(TimeUnit unit) {
```
```suggestion
    public long getDelay(@Nonnull TimeUnit unit) {
```

KarmaGYZ commented on flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/RpcEndpoint.java:
```diff
@@ -211,11 +231,29 @@ protected final void stop() {
         return stopFuture;
     }
 
+    protected void registerResource(Closeable closeableResource) {
```
Missing javadoc.

KarmaGYZ commented on flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/RpcEndpoint.java:
```diff
@@ -211,11 +231,29 @@ protected final void stop() {
+        }
+    }
+
+    protected boolean unregisterResource(Closeable closeableResource) {
```
Missing javadoc

KarmaGYZ commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java:
```diff
@@ -1811,6 +1834,12 @@ public void testJobMasterOnlyTerminatesAfterTheSchedulerHasClosed() throws Excep
+        assertJobMasterResourceClosed(jobMaster);
+    }
+
+    private void assertJobMasterResourceClosed(JobMaster jobMaster) {
```
I think we'd better test it in `RpcEndpointTest` and `FencedRpcEndpointTest`. It would be good to cover the scenario in which the fencing token changed.

KarmaGYZ reviewed (commented):
Thanks for the PR, @zjureel . It is already LGTM. I just left some minor comment with tests.

zjureel commented:
@KarmaGYZ Thanks for your comments, and i have fixed them

zjureel commented:
@flinkbot run azure

zjureel commented:
@flinkbot run azure

KarmaGYZ commented on flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/RpcEndpoint.java:
```diff
@@ -303,6 +355,11 @@ protected MainThreadExecutor getMainThreadExecutor() {
     }
 
+    @VisibleForTesting
+    public CloseableRegistry getResourceRegistry() {
```
```suggestion
    CloseableRegistry getResourceRegistry() {
```

KarmaGYZ reviewed (approved):
LGTM. @zjureel Please squash your commits and rebase it to the latest master.

@xintongsong Would you like to take another look?

zjureel commented:
Thanks @KarmaGYZ , I have done it according to your advice and updated this PR

xintongsong commented:
Thanks for working on this, @zjureel & @KarmaGYZ.

I'm still not entirely sure about exposing two executors (or two sets of scheduling interfaces) from the RPC endpoint. It's not easy to maintain as it requires callers to understand the differences between the two executors (tasks will be canceled as soon as the RPC endpoint shutdown for one executor, while not canceled immediately for the other).

I understand this is in response to the previous CI problems, where we suspect some clean-up operations are not completed due to tasks being canceled, and it's not easy to identify which clean-up operations are causing the problems. However, I'm not comfortable with scarifying the new design just to cover up for the real problems.

I'd suggest to try a few more things before we go for the two-executors approach.
- I noticed there are some calls to `schedule` with a zero delay, (e.g., `ApplicationDispatcherBootstrap#runApplicationAsync`). I think the purpose for these calls are to process asynchronously, rather than to schedule a future task. It might be helpful to check the delay value in `MainThreadExecutor#schedule`, and only use the scheduling thread pool when delay is positive. My gut feeling is most clean-up tasks are async tasks that needs to be executed immediately rather than in future.
- Another thing we may want to try is to replace `ExecutorService#shutdownNow` with `Executor#awaitTermination`, to leave a short time for the clean-up tasks to complete.

zjureel commented:
Thanks @xintongsong. Replace `ExecutorService#shutdownNow` with `Executor#awaitTermination` is not feasible, while the main target of `ExecutorService` is to clear the pending tasks when it closed. But it's a good idea to check the delay value in `MainThreadExecutor#schedule` before add it to `ExecutorService`, I'll try it. THX

zjureel commented:
@flinkbot run azure

zjureel commented:
@flinkbot run azure

XComp commented on flink-rpc/flink-rpc-core/src/test/java/org/apache/flink/runtime/concurrent/MainScheduledExecutorTest.java:
```diff
@@ -0,0 +1,151 @@
+    public void testScheduleRunnable() throws Exception {
+        MainScheduledExecutor mainScheduledExecutor =
+                new MainScheduledExecutor(new TestRunnableMainThreadExecutable());
+        final int timeDelay = 1;
```
1 second is quite a delay for such a test to succeed

XComp commented on flink-rpc/flink-rpc-core/src/test/java/org/apache/flink/runtime/concurrent/MainScheduledExecutorTest.java:
```diff
@@ -0,0 +1,151 @@
+import org.apache.flink.api.common.time.Time;
+import org.apache.flink.runtime.rpc.MainThreadExecutable;
+
+import org.junit.Test;
```
We might want to use JUnit5 here.

XComp commented on flink-rpc/flink-rpc-core/src/test/java/org/apache/flink/runtime/concurrent/MainScheduledExecutorTest.java:
```diff
@@ -0,0 +1,151 @@
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import static org.junit.Assert.assertEquals;
```
We might want to use AssertJ here.

XComp commented on flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/concurrent/MainScheduledExecutor.java:
```diff
@@ -0,0 +1,141 @@
+ * The main scheduled executor will manage the scheduled tasks. When the specified time arrives, the
+ * executor will send these tasks to gateway and execute them.
+ */
+public class MainScheduledExecutor implements ScheduledExecutor, Closeable {
```
What about calling it `OnMainThreadScheduledExecutor`? That would make the class name more descriptive.

XComp reviewed (commented):
Hi @zjureel, thanks for working on this issue. I was wondering what the state of this work is. The `scheduledExecutor` is something we are waiting for to implement [FLINK-25433](https://issues.apache.org/jira/browse/FLINK-25433). Our plan is to collect the cleanup tasks done by the Dispatcher in a single `ResourceCleaner` implementation (see [FLINK-25432](The `scheduledExecutor` is something we are waiting for to implement [FLINK-25433](https://issues.apache.org/jira/browse/FLINK-25432)). `FLINK-25433` would extend the functionality to allow periodic retries in case of failure. The initial intention was to provide some exponential delay for this as well. The `scheduledExecutor` integration you are working on in this PR comes in handy there.

I used the time going through the state of this PR to add some comments. All of them are of cosmetic nature. That leaves me with the question whether there's a plan to finalize your efforts? Thank you, Matthias~

KarmaGYZ commented:
Thanks for your review @XComp . These days, @zjureel and I do the investigation on whether this PR would cause FLINK-18356. As @xintongsong comments, we try to not expose two executors (or two sets of scheduling interfaces) from the RPC endpoint. However, that might compound the issue in FLINK-18356. At the moment, the table tests fail almost every time in our private azure, see 202220124.1 - 20220124.6 in https://dev.azure.com/karmagyz/Flink/_build?definitionId=1&_a=summary. It is actually blocking this PR. 

Once the table tests become stable, we will continue the progress of this PR.

XComp commented:
Thanks for the update @KarmaGYZ. I brought it up in today's release call mentioning that FLINK-25433 (which is a subtask of the efforts around introducing the `JobResultStore`; [FLIP-194](https://cwiki.apache.org/confluence/display/FLINK/FLIP-194%3A+Introduce+the+JobResultStore)) is blocked on this PR. But that said, I also underlined that FLINK-25433 is more like a nice-to-have feature as it's not blocking the integration of [FLIP-194](https://cwiki.apache.org/confluence/display/FLINK/FLIP-194%3A+Introduce+the+JobResultStore) in general.

dmvk commented:
Hi @zjureel, thanks for the PR. One question on the approach, why do we need to create a new implementation of the `ScheduledExecutor` instead of reusing the `ComponentMainThreadExecutor`, that already implements this interface?

KarmaGYZ commented:
@dmvk That's what we did in #18007. In this PR, we try to find which method has some specific logic and cannot use the new `schedule`. However, once we ensure that this PR is not related to FLINK-18356, we can return to the previous implementation.

zjureel commented:
Thanks @XComp @dmvk @KarmaGYZ , I have run this PR in my devops after rebase the commits from master in 20220207.1-20220207.8 in https://dev.azure.com/zjureel/flink/_build?definitionId=2, and I find that the issue in [FLINK-18356](https://issues.apache.org/jira/browse/FLINK-18356) has been fixed. As mentioned above I will return to the previous implementation in https://github.com/apache/flink/pull/18007 , what do you think? @KarmaGYZ  THX

zjureel commented:
@flinkbot run azure

zjureel commented:
Hi @KarmaGYZ @XComp  I have rebased the code from https://github.com/apache/flink/pull/18007 and updated junit4 to junt5 in RpcEndpointTest, FencedRpcEndpointTest and ScheduledFutureAdapterTest, could you review the codes again when you're free, thanks

KarmaGYZ commented on flink-rpc/flink-rpc-core/src/test/java/org/apache/flink/runtime/concurrent/ScheduledFutureAdapterTest.java:
```diff
@@ -18,94 +18,138 @@
 
 /** Unit tests for {@link ScheduledFutureAdapter}. */
-public class ScheduledFutureAdapterTest extends TestLogger {
+@ExtendWith(TestLoggerExtension.class)
```
The JUnit5 related refactor should belong to a separate hotfix commit.

KarmaGYZ commented on flink-runtime/src/test/java/org/apache/flink/runtime/rpc/FencedRpcEndpointTest.java:
```diff
@@ -38,24 +39,28 @@
 
-public class FencedRpcEndpointTest extends TestLogger {
+/** Tests for the FencedRpcEndpoint. */
+@ExtendWith(TestLoggerExtension.class)
```
The JUnit5 related refactor should belong to a separate hotfix commit.

KarmaGYZ commented on flink-runtime/src/test/java/org/apache/flink/runtime/rpc/RpcEndpointTest.java:
```diff
@@ -20,40 +20,48 @@
 
 /** Tests for the RpcEndpoint, its self gateways and MainThreadExecutor scheduling command. */
-public class RpcEndpointTest extends TestLogger {
+@ExtendWith(TestLoggerExtension.class)
```
The JUnit5 related refactor should belong to a separate hotfix commit.

KarmaGYZ commented on flink-runtime/src/test/java/org/apache/flink/runtime/rpc/RpcEndpointTest.java:
```diff
@@ -309,22 +354,83 @@ public void testScheduleCallableWithDelayInSeconds() throws Exception {
     private static void testScheduleWithDelay(
             BiConsumer<RpcEndpoint.MainThreadExecutor, Duration> scheduler) throws Exception {
-        final CompletableFuture<Long> actualDelayMsFuture = new CompletableFuture<>();
+        final CompletableFuture<Void> actualDelayMsFuture = new CompletableFuture<>();
```
Should we rename it to `taskCompletedFuture` or something else?

KarmaGYZ commented on flink-runtime/src/test/java/org/apache/flink/runtime/rpc/RpcEndpointTest.java:
```diff
@@ -309,22 +354,83 @@ public void testScheduleCallableWithDelayInSeconds() throws Exception {
+
+    private static void testScheduleAfterClose(
+            BiFunction<RpcEndpoint.MainThreadExecutor, Duration, ScheduledFuture<?>> scheduler) {
+        final CompletableFuture<Void> actualDelayMsFuture = new CompletableFuture<>();
```
Should we rename it to taskCompletedFuture or something else?

zjureel commented:
Thanks @KarmaGYZ , I've divided the commit into two commits with "update junit" and "add thread pool"

KarmaGYZ reviewed (approved):
LGTM. @xintongsong @XComp Would you like to take another look?

XComp commented:
Sorry with the delayed response. I'm busy with some other blocking PRs and might not be able to do another pass soon'ish. Hence, if somebody else is able to do a second pass, that'd be great.

xintongsong commented on flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/RpcEndpoint.java:
```diff
@@ -303,6 +348,11 @@ protected MainThreadExecutor getMainThreadExecutor() {
         return mainThreadExecutor;
     }
 
+    @VisibleForTesting
+    CloseableRegistry getResourceRegistry() {
+        return resourceRegistry;
+    }
```
I think the tests should not require exposing the closeable registry. Instead, they should validate that the executors are properly closed.

xintongsong commented on flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/RpcEndpoint.java:
```diff
@@ -211,11 +227,40 @@ protected final void stop() {
      */
     public final CompletableFuture<Void> internalCallOnStop() {
         validateRunsInMainThread();
+        try {
+            resourceRegistry.close();
+        } catch (IOException e) {
+            throw new RuntimeException("Close resource registry fail", e);
+        }
```
Instead of throwing a `RuntimeException`, I think the error should be reflected by complete the returning future exceptionally.

xintongsong reviewed (changes requested):
Thanks for working on this @zjureel, @KarmaGYZ. I have some more comments, please take a look.

zjureel commented on flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/concurrent/ComponentMainThreadExecutor.java:
```diff
@@ -30,11 +31,15 @@
  * org.apache.flink.runtime.rpc.RpcEndpoint}.
  */
-public interface ComponentMainThreadExecutor extends ScheduledExecutor {
+public interface ComponentMainThreadExecutor extends ScheduledExecutor, Closeable {
```
Sounds good to me, thanks

zjureel commented on flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/RpcEndpoint.java:
```diff
@@ -401,38 +451,90 @@ public void validateRunsInMainThread() {
+                                delayMillis,
+                                TimeUnit.MILLISECONDS);
+                return new ScheduledFutureAdapter<>(
+                        scheduledFuture, ft, delayMillis, TimeUnit.MILLISECONDS);
```
Cancel the `scheduledFuture` will remove the `command` from the queue of thread pool. As mentioned above, if the `command` won't be executed even later `gateway.scheduleRunAsync(ft, 0L)` is called after `ft` is canceled, it may no need to cancel the `scheduledFuture`. The queue of thread pool will be clear in `MainThreadExecutor.close` too, I will check and fix it

zjureel commented:
Hi @xintongsong , I have updated the codes. At present, the remaining point is the behavior after `mainScheduledExecutor` is shutdown. The `mainThreadExecutor` will execute the task with delay==0 even when the `mainScheduledExecutor` is shutdown. Could you help to review these codes when you're free? thanks

xintongsong commented on flink-runtime/src/test/java/org/apache/flink/runtime/rpc/RpcEndpointTest.java:
```diff
@@ -309,22 +354,83 @@ public void testScheduleCallableWithDelayInSeconds() throws Exception {
+        assertFalse(taskCompletedFuture.isDone());
+    }
+
+    private static void testCancelScheduledTask(
+            BiFunction<RpcEndpoint.MainThreadExecutor, Duration, ScheduledFuture<?>> scheduler) {
+        final CompletableFuture<Void> actualDelayMsFuture = new CompletableFuture<>();
+        final String endpointId = "foobar";
+
+        final MainThreadExecutable mainThreadExecutable =
+                new TestMainThreadExecutable((runnable) -> actualDelayMsFuture.complete(null));
+
+        final RpcEndpoint.MainThreadExecutor mainThreadExecutor =
+                new RpcEndpoint.MainThreadExecutor(mainThreadExecutable, () -> {}, endpointId);
+        final Duration timeDelay = Duration.ofSeconds(10);
+        ScheduledFuture<?> scheduledFuture = scheduler.apply(mainThreadExecutor, timeDelay);
+
+        assertEquals(1, mainThreadExecutor.getMainScheduledExecutor().getQueue().size());
+        scheduledFuture.cancel(true);
+        mainThreadExecutor.close();
+
+        assertTrue(scheduledFuture.isCancelled());
+        assertTrue(mainThreadExecutor.getMainScheduledExecutor().getQueue().isEmpty());
+        assertFalse(actualDelayMsFuture.isDone());
```
I see two problems in this test:

1. We should verify that `(runnable) -> actualDelayMsFuture.complete(null)` will not be executed, rather than verifying it's not passed to `mainThreadExecutor.runAsync()`. Actually, canceling `scheduledFuture` will not stop it from being scheduled to `mainThreadExecutor`. It is `FutureTask` that guarantees itself won't be executed once canceled.

Currently, the test will fail if you remove `mainThreadExecutor.close()`, because it is a `() -> {}` or `() -> 1` that actually wrapped int the `scheduledFuture`, thus cancelling it will not stop `actualDelayMsFuture` from being completed.

2. Another challenge here is to make sure task is not executed before being canceled.
- If we schedule with a short delay, the test might be unstable because the task can finish before canceling.
- If we schedule with a large delay, we would need to wait longer to see if the task is performed, if not looking into the internal queue of the executor.

I would suggest to use a short delay, while use `assumeTrue` to skip the test if the task is occasionally finished before canceling. You can find out whether the task has been finished before canceling from the return value of `cancel()`.

xintongsong commented on flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/RpcEndpoint.java:
```diff
@@ -401,38 +451,90 @@ public void validateRunsInMainThread() {
+                log.warn(
+                        "The scheduled executor service is shutdown and return throwing scheduled future for callable {}",
+                        callable);
+                return ThrowingScheduledFuture.getInstance();
```
@zjureel,

I think my major point was to figure out what was the previous behavior (when `schedule` is called on a closed endpoint) and to align with it. Looking at my previous comment now, I realize I haven't make that clear.
- If it is expected to fail explicitly, we should not rely on the returned future which can easily get ignored.
- If it is expected to be ignored, we won't need to return a special future.

I looked a bit more into how `schedule` behaves previously. It seems I was wrong about throwing a `RuntimeException`. In fact, commands scheduled on a closed endpoint will simply be ignored (with some info logs in `AkkaRpcActor`).

Therefore, I'd suggest not to introduce the special class `ThrowingScheduledFuture`, and simply log and ignore the command if `mainScheduledExecutor` is shutdown.

xintongsong commented:
@zjureel,
Thanks for addressing my comment. I've give this PR another pass. In addition to the `mainScheduledExecutor` shutdown behaviors, which I've replied, I have only one major concern regarding the implementation of the test case. Please take a look.

zjureel commented:
Thank you for your comments @xintongsong , ignore the tasks directly when the scheduled executor is shutdown is better, I have updated the test cases too.  Please review them again when you're free, thanks again :)

xintongsong reviewed (approved):
Thanks for addressing my comments, @zjureel. The PR LGTM now. +1 for merging.

KarmaGYZ commented:
@zjureel There is an `Architecture Violation` failure, please take a look.

BTW, please rebase it to the latest master as the FLINK-18356 has been fixed.

zjureel commented:
@flinkbot run azure
